aviatrix/resource_aviatrix_spoke_external_device_conn.go

```diff
@@ -670,6 +707,61 @@ func resourceAviatrixSpokeExternalDeviceConnCreate(d *schema.ResourceData, meta
 		return fmt.Errorf("failed to create Aviatrix external device connection: %s", err)
 	}
 
+	enableBFD, ok := d.Get("enable_bfd").(bool)
+	if !ok {
+		return fmt.Errorf("expected enable_bfd to be a boolean, but got %T", d.Get("enable_bfd"))
+	}
+	if enableBFD && externalDeviceConn.ConnectionType != "bgp" {
+		return fmt.Errorf("BFD is only supported for BGP connection type")
+	}
+	externalDeviceConn.EnableBfd = enableBFD
+	// set the bgp bfd config details only if the user has enabled BFD
+	if enableBFD {
+		bgp_bfd, ok := d.Get("bgp_bfd").([]interface{})
+		if !ok {
+			return fmt.Errorf("expected bgp_bfd to be a list of maps, but got %T", d.Get("bgp_bfd"))
+		}
+		// set bgp bfd using the config details provided by the user
+		if len(bgp_bfd) > 0 {
+			for _, bfd0 := range bgp_bfd {
+				bfd1, ok := bfd0.(map[string]interface{})
+				if !ok {
+					return fmt.Errorf("expected bgp_bfd to be a map, but got %T", bfd0)
+				}
+				transmitInterval := defaultBfdTransmitInterval
+				receiveInterval := defaultBfdReceiveInterval
+				multiplier := defaultBfdMultiplier
+				if value, ok := bfd1["transmit_interval"].(int); ok {
+					transmitInterval = value
+				}
+				if value, ok := bfd1["receive_interval"].(int); ok {
+					receiveInterval = value
+				}
+				if value, ok := bfd1["multiplier"].(int); ok {
+					multiplier = value
+				}
+				bfd2 := &goaviatrix.BgpBfdConfig{
+					TransmitInterval: transmitInterval,
+					ReceiveInterval:  receiveInterval,
+					Multiplier:       multiplier,
+				}
```

Comment:
Suggest factoring this into a function that takes the terraform config and returns the BgpBfdConfig struct, or error.

You ought to be able to write tests to exercise the error paths of that.

Reply:
There's also half a dozen other places where you could use this function.

Reply:
updated